server/status.go

```diff
@@ -152,19 +154,98 @@ func (s *statusServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	s.router.ServeHTTP(w, r)
 }
 
-// handleGossipStatus handles GET requests for gossip network status.
-func (s *statusServer) handleGossipStatus(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
-	w.Header().Set(util.ContentTypeHeader, util.JSONContentType)
+// extractNodeID examines the node_id URL parameter and returns the nodeID and a
+// boolean showing if it is this node. If node_id is "local" or not present, it
+// returns the local nodeID.
+func (s *statusServer) extractNodeID(ps httprouter.Params) (proto.NodeID, bool, error) {
+	nodeIDParam := ps.ByName("node_id")
+
+	// No parameter provided or set to local.
+	if len(nodeIDParam) == 0 || localRE.MatchString(nodeIDParam) {
+		return s.gossip.GetNodeID(), true, nil
+	}
+
+	id, err := strconv.ParseInt(nodeIDParam, 10, 64)
+	if err != nil {
+		return proto.NodeID(0), false, err
+	}
+	nodeID := proto.NodeID(id)
+	return nodeID, nodeID == s.gossip.GetNodeID(), nil
+}
+
+// proxy request performs a GET request to another node's status server.
+func (s *statusServer) proxyRequset(nodeID proto.NodeID, url string, w http.ResponseWriter, r *http.Request) {
+	addr, err := s.gossip.GetNodeIDAddress(nodeID)
+	if err != nil {
+		http.Error(w,
+			fmt.Sprintf("node could not be located: %s", nodeID),
+			http.StatusBadRequest)
+		return
+	}
+
+	// Create a call to the other node. We might want to consider moving this
+	// to an RPC instead of just proxying it.
+	// Generate the redirect url and copy all the parameters to it.
+	requestURL := fmt.Sprintf("%s://%s%s?%s", s.ctx.RequestScheme(), addr, url, r.URL.RawQuery)
+	req, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	resp, err := s.proxyClient.Do(req)
+	if err != nil {
+		log.Error(err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set(util.ContentTypeHeader, resp.Header.Get(util.ContentTypeHeader))
+
+	// Only pass through a whitelisted set of status codes.
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusNotFound, http.StatusBadRequest, http.StatusInternalServerError:
+		w.WriteHeader(resp.StatusCode)
+	default:
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	defer resp.Body.Close()
+	_, err = io.Copy(w, resp.Body)
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+// handleGossipLocal handles local requests for gossip network status.
+func (s *statusServer) handleGossipLocal(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
 	b, err := s.gossip.GetInfosAsJSON()
 	if err != nil {
 		log.Error(err)
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 	}
+	w.Header().Set(util.ContentTypeHeader, util.JSONContentType)
+	w.WriteHeader(http.StatusOK)
```

Comment:
No. It gets added automatically if we don't add another status and start writing.
I'll remove them.